docs/design/storage-auto-scale.md

```diff
@@ -0,0 +1,113 @@
+
+    b) If the expansion is not in progress set the status `phase` to `NotStarted` 
+
+2) List all the device class Osds used percentage using Prometheus metrics.
+    Query: (ceph_osd_metadata * on (ceph_daemon, namespace, managedBy) group_right(device_class,hostname) (ceph_osd_stat_bytes_used / ceph_osd_stat_bytes))
+               
+            OsdPercentage:
+                osd1-ssd               69%
+                osd2-ssd               72%
+
+3) If the highest osd percentage reported is more than osdScalingThresholdPercent(70%) means reaching osd nearfull, scaling is needed.
+
+4) If scaling is needed calculate the `expectedOsdSize` and `expectedOsdCount`.
+    1)  Calculate the total number osds and osd size from the prometheus response.
+    2)  Calculate the total number osds and osd size from the storageCluster.
+    3)  If the above values are equal(needed to avoid double re-sizing), create or update the `OsdPercentage map`
+        1)  If the Osd size is less than maxOsdSize(default:8Tib), do vertical scaling by doubling the each osd sizes for that device class.
+        2)  If the Osd size is equal to maxOsdSize(default:8Tib), do a horizontal scaling, by adding 1 osd of maxOsdSize(default:8Tib) on each `storageDeviceSet`.
+      
+            
+            `OsdPercentage` map:
+            ```
+                expectedOsdSize: 4
+                expectedOsdCount: 6
+            ```
+
+5) StorageAutoScaling status update:
+   1) Depending on the `OsdPercentage map` set the `expectedOsdSize` and `expectedOsdCount` to reflect the new expected value.
+   2) Update `phase` to `InProgress` on the `StorageAutoScaling` CR which need scaling.
+   3) Update the `lastExpansionStartTime` as `current-time`.
+
+6) Scale with the following algorithm:
+   1) Check if the totalCapacity allows to scale it or not.
+   2) Patch the Storagecluster, with all the device sets update needed at the same time.
+
+7) Verify and Alert:
+   1) Verify the Storagecluster whether the new osds are added or scaled in size, for all the device sets. 
+   2) If the scaling is successful will raise a `Succeeded` alert and update the status of the `StorageAutoScaling` CR with `lastExpansionCompletionTime` and `phase` and also osd count and size.
+   3) If it is fails will do a requeue every 1 min and, will raise a `failure` alert if scaling not `Succeeded` with in timeoutSeconds(default:1800) interval.   
+
+### InProgress state:
+
+Based on the above algorithm there would be two conditions where in-progress is set, elaborating those conditions,
+
+1) If scaling is just started:
+    1) Set `phase` to `InProgress`.
+    2) Verify is the scaling is successful. 
+    3) If the scaling is successful set the `phase` to `completed`.
+    4) If the scaling is un-successful requeue every 1 min, we have the 2nd case. 
+
+2) If the scaling has already started and its requeue
+    1) Now the requeue will happen every 1 min.
```

Comment:
@nb-ohad No concern with the re-queue pattern for checking status of an in-progress scaling, right?

Reply:
1. I have no issue with that, mainly because if we consider scaling in progress as an invalid state for acting on new configuration changes then this is just a plain error that will result in a re-queuing. 

2. Putting the above aside, I would like to understand what is that issue of acting on a change on the desired state while we are in progress?  k8s itself can handle it, so why can't we?

Reply:
what would be the source of truth for the desired state, we don't have desired state in the cr spec, and the sync map we even don't have the desired count, its moreover in the cr status that is been updated

Reply:
> would like to understand what is the issue of acting on a change on the desired state while we are in progress? k8s itself can handle it, so why can't we? 

The desired state is internally calculated by the controller so let's finish the in-progress state